packages/@aws-cdk/aws-codepipeline/lib/action.ts

```diff
@@ -111,20 +226,272 @@ export interface ActionBindOptions {
 }
 
 export interface ActionConfig {
+  /**
+   * Static configuration returned by bind()
+   *
+   * @deprecated Return boundAction instead.
+   */
   readonly configuration?: any;
+
+  /**
+   * A bound version of the action
+   *
+   * Provides access to methods that can only be called once the
+   * Action has been bound to a pipeline.
+   *
+   * If this value is set, `configuration` will be ignored.
+   */
+  readonly boundAction?: IBoundAction;
+}
+
+/**
+ * The result of binding an Action
+ */
+export interface IBoundAction {
```

Comment:
Ideally, we would have done `IBoundAction extends IAction` and you wouldn't have had this.

However, TypeScript's structural typing will fuck us over there: the natural implementation of that would have been:


```
class Action implements IAction {
  public bind(): IBoundAction {
    return this;
  }

  public render() {
  }
}
```

HOWEVER at this point, you can already pass `Action` into a place expecting an `IBoundAction`, even though you're supposed to call `bind()` first (to have this contract guarantee).

So a "parent pointer" is a cheap way to be able to access the original `IAction` interface without actually having `Action` accidentally implement `IBoundAction`. 

Honestly, I guess just having `actionProperties` on there would have been good enough as well, since we don't really care about the actual `Action` object, just about the `inputs`/`outputs`/etc.

Reply:
Sorry, I still don't understand.

There is basically a single place in the codebase that `bind()` is called, and it is the CodePipeline construct. Why do we need `boundAction` that points to the same `IAction` that we have available?